btree/btree.c

```diff
@@ -3152,13 +3152,31 @@ int64_t btree_create_tick(struct m0_sm_op *smop)
 int64_t btree_destroy_tick(struct m0_sm_op *smop)
 {
 	struct m0_btree_op 	*bop = M0_AMB(bop, smop, bo_op);
-	//ToDo: Implement complete destroy tick function.
+
 	switch(bop->bo_op.o_sm.sm_state)
 	{
 		case P_INIT:
+			M0_PRE(bop->bo_arbor != NULL);
+			M0_PRE(bop->bo_arbor->t_desc != NULL);
+			M0_PRE(node_invariant(bop->bo_arbor->t_desc->t_root));
+
+			/** The following pre-condition is currently a
+			 *  compulsion as the delete routine has not been
+			 *  implemented yet.
+			 *  Once it is implemented, this pre-condition can be
+			 *  modified to compulsorily remove the records and get
+			 *  the node count to 0.
+			 */
+			
+			M0_PRE(node_count(bop->bo_arbor->t_desc->t_root) == 0);
+
+			tree_put(bop->bo_arbor->t_desc);
 			return P_ACT;
 
 		case P_ACT:
+			m0_free(bop->bo_arbor);
```

Comment:
Practically there is no need to have this separate state since this state only frees up memory which can be done in the previous state. If there is any operation which seems to take long time then we need to try to return back to the caller before moving to the next state.

Reply:
Updated in the latest commit.